File: src/anim.c
/* ------------------------------------------------------------------------
 * anim.c - the animation gallery.  One effect, in isolation, on demand.
 *
 * Every moving thing in this game is wired into a moment: the curtain only
 * opens on the way to the menu, the trophy is only poured when somebody
 * wins, and the laser only cuts when a piece is taken.  That makes them
 * expensive to look at and almost impossible to compare - to see the
 * tumble you have to play a turn, and to see the trophy you have to play a
 * match.
 *
 * This runs one of them, alone, against a plain background, and parks.
 * Which one is chosen at compile time with -DANIM=n, so each is a separate
 * binary and a screenshot of it is deterministic: no game state, no
 * randomness that matters, nothing else on screen to confuse what changed.
 *
 *     make anim              the gallery, as one contact sheet
 *     make anim-run ANIM=3   one of them, in the emulator, to watch
 *
 * The numbering is in tools/anim.py, which is also what builds the sheet.
 * --------------------------------------------------------------------- */

#include "plus4.h"
#include "rules.h"
#include "board.h"
#include "text.h"
#include "dice.h"
#include "front.h"
#include "music.h"
#include "etch.h"

#ifndef ANIM
#  define ANIM 0
#endif

/* The board, dressed enough that the effects which sit on it have
** something to sit on.  Several of them - the shimmer, the flash, the
** glide - are meaningless against an empty screen, because what they do is
** modify a board that is already there. */
static void with_board (void)
{
    board_init ();
    shade_profile (SHADE_SOFT);
    reset_board ();
    board_draw ();
}

static void label (const char* s)
{
    text_put (1, 24, s, CBYTE (5, 1));
}

void main (void)
{
    unsigned char i;

    board_init ();
    shade_profile (SHADE_SOFT);
    music_init ();
    TED_BGCOLOR = CBYTE (0, 0);
    TED_BORDER  = CBYTE (0, 0);
    screen_fill (CH_SPACE, 0);

#if ANIM == 0
    /* THE LASER, cutting a word.  What a capture shows. */
    label ("0  laser etch - text");
    for (;;) {
        etch_text ("captured", 12, 100);
        wait_frames_live (60);
        rows_fill (10, 14, CH_SPACE, 0);
    }

#elif ANIM == 1
    /* THE LASER, cutting the two letters from a bitmap.  The way in. */
    label ("1  laser etch - big ur");
    for (;;) {
        etch_big_ur (250);
        wait_frames_live (80);
        rows_fill (8, 18, CH_SPACE, 0);
    }

#elif ANIM == 2
    /* THE SHIMMER.  A sheen drifting along each tile's face - the outer
    ** bands leftward, the middle band rightward - every tile out of phase
    ** with its neighbours.  Needs the board: it modifies the highlight and
    ** shadow rows of tiles that are already drawn. */
    with_board ();
    label ("2  shimmer - drift");
    for (;;) wait_frames_live (1);

#elif ANIM == 3
    /* THE TURN SWEEP.  The flare that marks the players changing hands. */
    with_board ();
    label ("3  turn sweep");
    for (;;) {
        for (i = 0; i < 6; ++i) {
            TED_BGCOLOR = CBYTE ((unsigned char)(7 - i), 9);
            TED_BORDER  = CBYTE ((unsigned char)(i + 2), 9);
            wait_frames_live (6);
        }
        TED_BGCOLOR = CBYTE (0, 0);
        TED_BORDER  = CBYTE (0, 0);
        wait_frames_live (40);
    }

#elif ANIM == 4
    /* THE TUMBLE.  Four tetrahedral lots scattered across the floor and
    ** settling one at a time. */
    with_board ();
    label ("4  the lots tumble");
    thrower_colour = CBYTE (6, 9);
    for (;;) {
        (void)throw_lots ();
        wait_frames_live (60);
        floor_wipe ();
    }

#elif ANIM == 5
    /* THE CURTAIN, and the lamp chase that runs while it moves. */
    label ("5  curtain and lamps");
    for (;;) {
        theatre_draw (CURTAIN_SHUT);
        curtain_open (0);
        for (i = 0; i < 60; ++i) { lamps_frame (i); wait_frames_live (2); }
        curtain_close ();
    }

#elif ANIM == 6
    /* THE TROPHY, poured a row at a time with a name cut into the bowl. */
    label ("6  the trophy");
    for (;;) {
        screen_fill (CH_SPACE, 0);
        label ("6  the trophy");
        trophy_draw ("urbot");
        wait_frames_live (80);
    }

#elif ANIM == 7
    /* THE GLIDE.  A piece travelling one character cell at a time along
    ** the line between two squares, rather than jumping a square. */
    with_board ();
    label ("7  piece glide");
    for (;;) {
        glide (piece_cell_x (0, 1), piece_cell_y (0, 1),
               piece_cell_x (0, 4), piece_cell_y (0, 4),
               (unsigned char)(CH_PIECE_BASE + 1), CBYTE (6, 9), 2);
        wait_frames_live (25);
        board_draw ();
        label ("7  piece glide");
    }

#elif ANIM == 8
    /* THE MENU SCROLL.  The choices climbing into the open stage. */
    label ("8  menu scroll-in");
    for (;;) {
        theatre_draw (CURTAIN_OPEN);
        for (i = 11; i > 0; --i) {
            rows_fill (12, 22, CH_SPACE, 0);
            text_centre ((unsigned char)(12 + i - 1), "player vs urbot",
                         CBYTE (7, 1));
            text_centre ((unsigned char)(13 + i - 1), "two players",
                         CBYTE (7, 1));
            wait_frames_live (5);
        }
        wait_frames_live (60);
    }

#elif ANIM == 10
    /* THE BURN, on the whole stage.  What a capture shows: a bed of fire
    ** across the bottom six rows and the word in block letters above it,
    ** catching from its feet upward, glyphs morphing through flame rather
    ** than colours changing under a fixed character.  Against the real
    ** board, because the apron it takes over is the board's own floor. */
    /* THE CHRONICLE IS THE TEST, and it is written here rather than a
    ** label being put at the foot of the screen.  The fire covers the log
    ** as well as the floor, so the thing most likely to be wrong is not
    ** the fire - it is whether four lines of words come back afterwards.
    ** They are in the chronicle's own buffer and apron_clear repaints
    ** them; a shot taken in the pause after the fire says whether that is
    ** true. */
    with_board ();
    say ("10  burn - the apron alight");
    say ("the fire takes these four lines too");
    say ("urbot a captures a foe on 8");
    say ("...and the log must come back");
    for (;;) {
        burn_apron ("captured", 190);
        wait_frames_live (40);
        apron_clear ();
        wait_frames_live (120);         /* long enough to be caught       */
    }

#elif ANIM == 12
    /* THE WHOLE END OF A MATCH, in the order victory() does it and with
    ** the same numbers.
    **
    ** Every other entry here shows ONE effect; this one shows a SEQUENCE,
    ** because the things that go wrong at the end of a match are not
    ** inside any single effect - they are between them. Does the cup land
    ** on top of the name. Does the wipe between them take the two lines
    ** above it as well. Does the border come back to the winner's colour
    ** when the last shell dies, or stay stuck on a firework's hue.
    **
    ** None of that is reachable from the game without playing a match to
    ** a win, which is why it is here: a screenshot of the real victory is
    ** a screenshot of a demo that happened to be finishing. */
    TED_BORDER = CBYTE (6, 9);          /* stand in for the winner's colour */
    text_centre (13, "urfinkel wins the royal game of ur", CBYTE (7, 7));
    text_centre (14, "the other side got 3 home", CBYTE (5, 7));
    for (;;) {
        etch_word ("urfinkel", 16, 200);
        wait_frames_live (60);
        rows_fill (APRON_TOP + 1, APRON_BOT, CH_SPACE, 0);
        trophy_draw ("urfinkel");
        fireworks (250);
        wait_frames_live (60);
        rows_fill (APRON_TOP + 1, APRON_BOT, CH_SPACE, 0);
    }

#elif ANIM == 9
    /* THE FIREWORKS.  Three shells in the air at once over a name, as the
    ** winner gets, and the border lighting up on every burst. */
    label ("9  fireworks");
    for (;;) {
        etch_word ("urbot", 16, 90);
        fireworks (250);
        wait_frames_live (30);
        rows_fill (4, 22, CH_SPACE, 0);
        label ("9  fireworks");
    }

#elif ANIM == 11
    /* THE WINNER'S NAME, cut across the whole stage in block letters -
    ** eight characters at a five column pitch is thirty-nine of the
    ** forty. */
    with_board ();
    label ("11  block letters - the win");
    for (;;) {
        etch_word ("urfinkel", 16, 200);
        wait_frames_live (60);
        apron_clear ();
        label ("11  block letters - the win");
        wait_frames_live (30);
    }

#else
    label ("no such animation");
    for (;;) wait_frames_live (1);
#endif
}

File: anim.py
#!/usr/bin/env python3
"""Build the animation gallery: every effect, caught mid-motion, on one sheet.

Each effect is a separate binary (src/anim.c, -DANIM=n) running that one
thing on a loop against a plain background.  This builds them all, shoots
each at a cycle count chosen to land INSIDE the motion rather than at
either end of it, and lays the results out as a contact sheet.

Why a separate binary each: a screenshot of an effect embedded in the game
is a screenshot of the game, and what it caught depends on the match being
played.  One effect alone, looping, with nothing else on screen, is
reproducible - shoot it twice and get the same picture.

Two shots per effect, a little apart, because a still frame cannot show
whether something is moving.  The pair is the evidence.

    make anim
"""

import os
import subprocess
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
ROOT = os.path.dirname(HERE)
BUILD = os.path.join(ROOT, "build")
BREW = os.environ.get("BREW_PREFIX", "/opt/homebrew")

CORE = ["board.c", "rules.c", "text.c", "dice.c", "urbot.c", "music.c",
        "front.c", "kbd.c", "etch.c", "dbg.c", "blit.s", "irq.s"]

# n, title, and the two cycle counts to shoot at.  The counts are chosen
# per effect because they run at wildly different lengths - the trophy is
# poured over four seconds, the turn sweep is over in under one.
#
# NOTHING BEFORE ABOUT 110e6, AND THAT NUMBER MOVES WITH THE BINARY.  A
# shot before the PRG has finished loading catches the LOADING screen and
# viceshot rejects it, which is what a "?" in the run output means.  These
# were written when the binaries were thirty kilobytes and the threshold
# was 60e6; the transcribed beds took them past fifty-five kilobytes and
# the whole sheet quietly became twelve pictures of a loader - every row
# "?" - until 2026-08-08.  Measured then by bisection on anim2, which
# draws the board the instant it starts: still loading at 105e6, running
# at 110e6.  If it happens again the fix is here and not in the effects -
# bisect the same way and move every number below by the difference.
#
# The second shot of each pair is deliberately far from the first, because
# a still cannot show motion and the pair is the evidence.
SHOTS = [
    (0, "laser etch - text",      (119_000_000, 123_000_000)),
    (1, "laser etch - big UR",    (115_000_000, 125_000_000)),
    (2, "shimmer - orbiting",     (120_000_000, 121_500_000)),
    (3, "turn sweep",             (118_000_000, 119_500_000)),
    (4, "the lots tumble",        (119_000_000, 121_000_000)),
    (5, "curtain and lamps",      (120_000_000, 135_000_000)),
    (6, "the trophy",             (115_000_000, 125_000_000)),
    (7, "piece glide",            (118_000_000, 119_000_000)),
    (8, "menu scroll-in",         (115_000_000, 121_000_000)),
    # The fireworks only start after the name has been cut, which is 90
    # frames in - hence much later than the rest.
    (9, "fireworks",              (262_600_000, 263_400_000)),
    # The two that take the whole apron.  Both are long - a capture runs
    # 190 frames and a block-letter cut 200 - so the pair is shot wide
    # apart: early enough to catch the fire building and the first letters
    # being cut, late enough to catch the blaze and the finished word.
    (10, "burn - the apron alight", (150_000_000, 200_000_000)),
    (11, "block letters - the win", (175_000_000, 190_000_000)),
    # Not one effect but the whole end of a match, in victory()'s own
    # order and with its own numbers.  What goes wrong at the end of a
    # match is BETWEEN the effects - the cup landing on the name, the wipe
    # eating the two lines above it, the border left on a firework's hue -
    # and no single-effect entry can show any of that.
    (12, "the end of a match",     (185_000_000, 235_000_000)),
]


def build(n):
    prg = os.path.join(BUILD, "anim%d.prg" % n)
    cmd = ([os.path.join(BREW, "bin", "cl65"), "-t", "plus4", "-Osir", "-Cl",
            "-DBUILD_DATE=\"gallery\"", "-DANIM=%d" % n, "-o", prg,
            os.path.join(ROOT, "src", "anim.c")]
           + [os.path.join(ROOT, "src", f) for f in CORE])
    r = subprocess.run(cmd, capture_output=True, text=True, cwd=ROOT)
    if r.returncode:
        print(r.stdout + r.stderr, file=sys.stderr)
        raise SystemExit("anim%d failed to build" % n)
    return prg


def shoot(prg, png, cycles):
    r = subprocess.run(
        [sys.executable, os.path.join(HERE, "viceshot.py"), prg, png,
         str(cycles), "--max-white=0.25"],
        capture_output=True, text=True, cwd=ROOT,
        env=dict(os.environ, BREW_PREFIX=BREW),
    )
    return r.returncode == 0 and os.path.exists(png)


def sheet(rows):
    from PIL import Image, ImageDraw
    if not rows:
        raise SystemExit("nothing shot")
    w, h = Image.open(rows[0][1]).size
    scale = 2
    tw, th = w // scale, h // scale
    pad, lab = 6, 14
    cols = 2
    n = len(rows)
    gw = cols * (tw * 2 + pad * 3)
    gh = ((n + cols - 1) // cols) * (th + lab + pad) + pad
    out = Image.new("RGB", (gw, gh), (24, 24, 28))
    d = ImageDraw.Draw(out)
    for i, (title, a, b) in enumerate(rows):
        cx = (i % cols) * (tw * 2 + pad * 3) + pad
        cy = (i // cols) * (th + lab + pad) + pad
        d.text((cx, cy), title, fill=(230, 230, 210))
        for j, p in enumerate((a, b)):
            im = Image.open(p).convert("RGB").resize((tw, th), Image.NEAREST)
            out.paste(im, (cx + j * (tw + pad), cy + lab))
    path = os.path.join(BUILD, "anim-sheet.png")
    out.save(path)
    return path


def main():
    os.makedirs(BUILD, exist_ok=True)
    rows = []
    for n, title, (c1, c2) in SHOTS:
        prg = build(n)
        a = os.path.join(BUILD, "anim%d-a.png" % n)
        b = os.path.join(BUILD, "anim%d-b.png" % n)
        ok_a, ok_b = shoot(prg, a, c1), shoot(prg, b, c2)
        moving = "?"
        if ok_a and ok_b:
            moving = "moving" if open(a, "rb").read() != open(b, "rb").read() \
                     else "STILL"
        print("  %d  %-24s %s" % (n, title, moving))
        if ok_a and ok_b:
            rows.append(("%d  %s  (%s)" % (n, title, moving), a, b))
    print()
    print("sheet:", sheet(rows))


if __name__ == "__main__":
    main()
